### DISCUSSION TITLE
Can firejail sandbox virt-manager and all other related processes?

### DISCUSSION THREAD

**Discussion by denizakcal** [netblue30/firejail#4526] [Q&A] [answered]
Hello.

I made a profile and local file for virt-manager ( https://packages.debian.org/bullseye/virt-manager ) (by modifying what firejail --build created).

When I do firejail --list after running firejail virt-manager and entering the root password to be able to connect to the VMs, I get
353617:user::firejail virt-manager

, as expected.

Then, when I close virt-manager, ps -e | grep -i virt shows that libvirtd (and sometimes also virtlogd) are running (which is something that happens whether I use firejail or not), but firejail --list doesn’t show anything.

So, I’m wondering, is it only the virt-manager GUI that's being sandboxed (and not the processes that do the actual virtualizing)?

If so, why isn’t the libvirtd (and possibly also virtlogd) being sandboxed?

If not, then why is firejail --list not showing that firejail virt-manager is still running or that a child process of it is?

It seems to me like libvirtd and virtlogd are sibling processes or something and that they’re not part of the sandbox.

Does anyone have an explanation as to what is happening?

Is this a bug?

Any input that could potentially help me get to the bottom of this would be greatly appreciated!

**Comment by rusty-snake** [comment, marked as answer]
> I made a profile and local file for virt-manager

It's always good to post a profile so everyone know your sandbox configuration that can affect a lot of questions/answers.

> So, I’m wondering, is it only the virt-manager GUI that's being sandboxed (and not the processes that do the actual virtualizing)?

Yes

> If so, why isn’t the libvirtd (and possibly also virtlogd) being sandboxed?

Because they aren't configured to be executed in firejail.

> If not, then why is firejail --list not showing that firejail virt-manager is still running or that a child process of it is?

If you close virt-manager (and all it's children) that what to expect.

> Does anyone have an explanation as to what is happening?

virt-manager/virsh/gnome-boxes are only UIs for libvirt which is started by systemd via socket-based activation.

**TL;DR:** Sandbox libvirtd with systemd's native features, see https://github.com/netblue30/firejail/wiki/Comparison-of-firejail-and-systemd's-hardening-options for examples.

> Is this a bug?

It's a feature :wink:

**Reply by denizakcal** [reply]
Sorry for the delay, and thanks for the response!

Thanks to the link you sent that maps Firejail syntax to its similar or equivalent Systemd version, the actual syntax in security.conf file (see below) shouldn't be a problem for me to complete (whenever I get around to it, in the near-ish future).

To simplify for the context of this post, though, just imagine the security.conf file is the following.:
```
[Service]
PrivateTmp=yes
```

So, basically, I typed out the following commands, but instead of pressing enter, I pressed the Tab key twice to bring about the autocomplete suggestions so that I can figure out all the Systemd units I need to sandbox (which are all the Systemd units related to virtualization).
```
root@debian:~# systemctl edit virt
virt-guest-shutdown.target  virtlockd.service           virtlogd-admin.socket       virtlogd.socket             
virtlockd-admin.socket      virtlockd.socket            virtlogd.service            
root@debian:~# systemctl edit libvirt
libvirtd-admin.socket   libvirtd.service        libvirtd-tcp.socket     libvirt-guests.service  
libvirtd-ro.socket      libvirtd.socket         libvirtd-tls.socket     
root@debian:~#
```

So, I issued the following commands, which create the directories for the Systemd units and then create a symlink from all of the other aforementioned directories to the (one) security.conf in /etc/systemd/system/libvirtd.service.d). (I did this because I think all the virtualization-related Systemd units need the exact same Systemd syntax for the sandboxing, and given that assumption, it's better to have one file to keep track of.):
```
mkdir -p /etc/systemd/system/libvirtd.service.d
mkdir -p /etc/systemd/system/libvirtd-admin.socket.d
mkdir -p /etc/systemd/system/libvirtd-ro.socket.d
mkdir -p /etc/systemd/system/libvirtd.service.d
mkdir -p /etc/systemd/system/libvirtd.socket.d
mkdir -p /etc/systemd/system/libvirtd-tcp.socket.d
mkdir -p /etc/systemd/system/libvirtd-tls.socket.d
mkdir -p /etc/systemd/system/libvirt-guests.service.d
mkdir -p /etc/systemd/system/virt-guest-shutdown.target.d
mkdir -p /etc/systemd/system/virtlockd-admin.socket.d
mkdir -p /etc/systemd/system/virtlockd.service.d
mkdir -p /etc/systemd/system/virtlockd.socket.d
mkdir -p /etc/systemd/system/virtlogd-admin.socket.d
mkdir -p /etc/systemd/system/virtlogd.service.d
mkdir -p /etc/systemd/system/virtlogd.socket.d

ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/libvirtd-admin.socket.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/libvirtd-ro.socket.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/libvirtd.service.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/libvirtd.socket.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/libvirtd-tcp.socket.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/libvirtd-tls.socket.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/libvirt-guests.service.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/virt-guest-shutdown.target.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/virtlockd-admin.socket.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/virtlockd.service.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/virtlockd.socket.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/virtlogd-admin.socket.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/virtlogd.service.d
ln -s  /etc/systemd/system/libvirtd.service.d/security.conf /etc/systemd/system/virtlogd.socket.d
```

**Could you please let me know if that was what I was supposed to do** (as far as you know, at least)**?** (I ask because I don't know much about Systemd, and I don't want to have a false sense of security.)

**Reply by rusty-snake** [reply]
This looks right so far for the service units. You do not need/can sandbox socket units.

Helpful commands:
- `systemctl cat libvirtd.service`
- `systemctl show libvirtd.service`
- `systemctl list-units`
- `systemctl restart libvirtd.service`
- `systemctl daemon-reload`

**Comment by denizakcal** [comment]
Thanks again! (And, sorry for the delay again!)
